compiler/optimizer/src/main/java/org/apache/nemo/compiler/optimizer/pass/compiletime/reshaping/IntermediateAccumulatorInsertionPass.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.nemo.compiler.optimizer.pass.compiletime.reshaping;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.nemo.common.Util;
+import org.apache.nemo.common.exception.SchedulingException;
+import org.apache.nemo.common.ir.IRDAG;
+import org.apache.nemo.common.ir.edge.IREdge;
+import org.apache.nemo.common.ir.edge.executionproperty.CommunicationPatternProperty;
+import org.apache.nemo.common.ir.vertex.OperatorVertex;
+import org.apache.nemo.common.ir.vertex.executionproperty.ParallelismProperty;
+import org.apache.nemo.common.ir.vertex.executionproperty.ShuffleExecutorSetProperty;
+import org.apache.nemo.common.ir.vertex.utility.IntermediateAccumulatorVertex;
+import org.apache.nemo.compiler.frontend.beam.transform.CombineTransform;
+import org.apache.nemo.compiler.optimizer.pass.compiletime.Requires;
+
+import java.io.File;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Pass for inserting intermediate aggregator for partial shuffle.
+ */
+@Requires(ParallelismProperty.class)
+public class IntermediateAccumulatorInsertionPass extends ReshapingPass {
+  private final String networkFilePath;
+  private boolean isUnitTest = false;
+  private static final Map<String, ArrayList<String>> UNIT_TEST_NETWORK_FILE = getUnitTestNetworkFile();
+
+  /**
+   * Default constructor.
+   */
+  public IntermediateAccumulatorInsertionPass() {
+    super(IntermediateAccumulatorInsertionPass.class);
+    this.networkFilePath = Util.fetchProjectRootPath() + "/tools/network_profiling/labeldict.json";
+  }
+
+  /**
+   * Constructor for unit test.
+   * @param isUnitTest indicates unit test.
+   */
+  public IntermediateAccumulatorInsertionPass(final boolean isUnitTest) {
+    this();
+    this.isUnitTest = isUnitTest;
+  }
+
+  private static Map<String, ArrayList<String>> getUnitTestNetworkFile() {
```

Comment:
Adding unit test codes here doesn't look good to me.

Reply:
It's been moved to a resources file.